{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7a0c3548-609d-4a6b-be4b-aed60a497628",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SVM Predicted labels: [0 0 0 1 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 0 0 1 1 1 1 0 0 1 0 0 1 1\n",
      " 0 0 0 1 0 1 1 0 1 0 1 1 0 0 1 0 0 0 1 0 1 0 0 1 0 1 1 0 1 0 1 0 0 0 0 1 0\n",
      " 1 0]\n",
      "SVM Accuracy: 0.8421052631578947\n",
      "SVM Precision: 0.9310344827586207\n",
      "SVM Sensitivity (Recall): 0.7297297297297297\n",
      "SVM F1 Score: 0.8181818181818181\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/biolab/.local/lib/python3.8/site-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from sklearn.feature_extraction import DictVectorizer\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.svm import SVC\n",
    "import numpy as np\n",
    "\n",
    "def KMer(text, k):\n",
    "    kmers_count = {}\n",
    "    s = 0\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i + k]\n",
    "        s += 1\n",
    "        if kmer in kmers_count:\n",
    "            kmers_count[kmer] += 1\n",
    "        else:\n",
    "            kmers_count[kmer] = 1\n",
    "    for key, value in kmers_count.items():\n",
    "        kmers_count[key] = value / s\n",
    "\n",
    "    return kmers_count\n",
    "\n",
    "def codon(text, k):\n",
    "    codon_count = {}\n",
    "    s = 0\n",
    "    i = 0\n",
    "    while i < len(text) - (k * 3) + 1:\n",
    "        kmer = text[i:i + (k * 3)]\n",
    "        s += 1\n",
    "        if kmer in codon_count:\n",
    "            codon_count[kmer] += 1\n",
    "        else:\n",
    "            codon_count[kmer] = 1\n",
    "        i += 3\n",
    "\n",
    "    for key, value in codon_count.items():\n",
    "        codon_count[key] = value / s\n",
    "\n",
    "    return codon_count\n",
    "\n",
    "def read_fasta_file(file_path):\n",
    "    sequences = []\n",
    "    current_sequence = None\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if current_sequence:\n",
    "                    sequences.append(current_sequence)\n",
    "                sequence_id = line[1:]\n",
    "                current_sequence = {'id': sequence_id, 'sequence': ''}\n",
    "            else:\n",
    "                current_sequence['sequence'] += line\n",
    "        if current_sequence:\n",
    "            sequences.append(current_sequence)\n",
    "    return sequences\n",
    "\n",
    "file_path = 'Arabidopsis_thaliana_BHLH_gene_Family.fasta'\n",
    "fasta = read_fasta_file(file_path)\n",
    "\n",
    "ids = []\n",
    "for id in fasta:\n",
    "    ids.append(id['id'])\n",
    "\n",
    "sequences = []\n",
    "y = []\n",
    "for seq in fasta:\n",
    "    sequences.append(seq['sequence'])\n",
    "    y.append(1)\n",
    "\n",
    "file_path = 'Arabidopsis_thaliana_CYP_gene_Family.fa'\n",
    "fasta = read_fasta_file(file_path)\n",
    "\n",
    "for seq in fasta:\n",
    "    sequences.append(seq['sequence'])\n",
    "    y.append(0)\n",
    "\n",
    "for id in fasta:\n",
    "    ids.append(id['id'])\n",
    "\n",
    "kmer_result = []\n",
    "for i in range(len(sequences)):\n",
    "    kmer_result.append(codon(sequences[i], 2))\n",
    "    kmer_result[i].update(codon(sequences[i], 1))\n",
    "\n",
    "\n",
    "v = DictVectorizer(sparse=False)\n",
    "\n",
    "features = v.fit_transform(kmer_result)\n",
    "feature_names = v.get_feature_names_out()\n",
    "\n",
    "\n",
    "x = pd.DataFrame(features, columns=feature_names)\n",
    "\n",
    "\n",
    "features = v.fit_transform(kmer_result)\n",
    "feature_names = v.get_feature_names_out()\n",
    "\n",
    "\n",
    "x.to_csv('kmer_features.csv', index=False)\n",
    "x.rename(index=dict(zip(x.index, ids)), inplace=True)\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)\n",
    "\n",
    "label_encoder = LabelEncoder()\n",
    "\n",
    "for col in x.columns:\n",
    "    if x[col].dtype == 'object':\n",
    "        x[col] = label_encoder.fit_transform(x[col])\n",
    "        \n",
    "y_train = label_encoder.fit_transform(y_train)\n",
    "y_test = label_encoder.transform(y_test)\n",
    "\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)\n",
    "X_test = np.array(X_test)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "scaler = StandardScaler()\n",
    "X_train_scaled = scaler.fit_transform(X_train)\n",
    "X_test_scaled = scaler.transform(X_test)\n",
    "svm_classifier = SVC(kernel='rbf', C=1)\n",
    "\n",
    "\n",
    "svm_classifier.fit(X_train_scaled, y_train)\n",
    "y_pred_svm = svm_classifier.predict(X_test_scaled)\n",
    "\n",
    "y_pred_svm_labels = label_encoder.inverse_transform(y_pred_svm)\n",
    "\n",
    "print(\"SVM Predicted labels:\", y_pred_svm_labels)\n",
    "\n",
    "\n",
    "accuracy_svm = accuracy_score(y_test, y_pred_svm)\n",
    "print(\"SVM Accuracy:\", accuracy_svm)\n",
    "\n",
    "precision_svm = precision_score(y_test, y_pred_svm)\n",
    "print(\"SVM Precision:\", precision_svm)\n",
    "\n",
    "sensitivity_svm = recall_score(y_test, y_pred_svm)\n",
    "print(\"SVM Sensitivity (Recall):\", sensitivity_svm)\n",
    "\n",
    "f1_svm = f1_score(y_test, y_pred_svm)\n",
    "print(\"SVM F1 Score:\", f1_svm)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03384e49-acfe-4dbb-92a6-09a0423bb4a5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
